### PR TITLE
Fix implicit account creation address sync for non Ed25519 addresses

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Fix implicit account creation address sync for non Ed25519 addresses

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Synced a wallet with an account address

### CODE REVIEW

File: sdk/src/wallet/operations/syncing/mod.rs [hunk 1]
```diff
@@ -99,16 +99,30 @@ where
...
         };
 
-        let address_to_sync = vec![
-            wallet_address_with_unspent_outputs,
-            AddressWithUnspentOutputs {
-                address: self.implicit_account_creation_address().await?,
-                output_ids: vec![],
-            },
-        ];
+        let mut addresses_to_sync = vec![wallet_address_with_unspent_outputs];
```
Comment (thibault-martinez) [inline sdk/src/wallet/operations/syncing/mod.rs:102]:
```suggestion
        let mut addresses_to_sync = vec![wallet_address_with_unspent_outputs];

```

File: sdk/src/wallet/operations/syncing/mod.rs [hunk 2]
```diff
@@ -99,16 +99,31 @@ where
...
+            if let Ok(implicit_account_creation_address) = self.implicit_account_creation_address().await {
+                addresses_to_sync.push(AddressWithUnspentOutputs {
+                    address: implicit_account_creation_address,
+                    output_ids: self
+                        .ledger()
+                        .await
+                        .unspent_outputs
+                        .values()
+                        .filter_map(|output_data| {
+                            if output_data.output.is_implicit_account() {
```
Comment (thibault-martinez) [inline sdk/src/wallet/operations/syncing/mod.rs]: I know we can only have one implicit address (for now?) but don't we want to actually check the address?

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs]: Didn't add it because currently it can't come form another address, but can add it if you want

Comment (thibault-martinez) [reply sdk/src/wallet/operations/syncing/mod.rs]: I think for completeness/futureproofing it would be nice yes

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs]: Added

File: sdk/src/wallet/operations/syncing/mod.rs [hunk 3]
```diff
@@ -94,21 +94,40 @@ where
...
+                addresses_to_sync.push(AddressWithUnspentOutputs {
+                    address: implicit_account_creation_address,
+                    output_ids: self
+                        .ledger()
+                        .await
+                        .unspent_outputs
+                        .values()
+                        .filter_map(|output_data| {
+                            if output_data.output.is_implicit_account()
+                                && output_data.output.as_basic().address() == wallet_address.inner()
```
Comment (thibault-martinez) [inline sdk/src/wallet/operations/syncing/mod.rs]: Shouldn't it be `implicit_account_creation_address`?

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs]: Yes, good catch

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs]: Fixed

**Review by thibault-martinez — DISMISSED** [review]

File: sdk/src/wallet/operations/syncing/mod.rs [hunk 4]
```diff
@@ -99,16 +99,33 @@ where
...
-                output_ids: vec![],
-            },
-        ];
+        let mut addresses_to_sync = vec![wallet_address_with_unspent_outputs];
+
+        if options.sync_implicit_accounts {
+            if let Ok(implicit_account_creation_address) = self.implicit_account_creation_address().await {
+                addresses_to_sync.push(AddressWithUnspentOutputs {
+                    output_ids: self
+                        .ledger()
```
Comment (DaughterOfMars) [inline sdk/src/wallet/operations/syncing/mod.rs:108]: call `Wallet::implicit_accounts` and filter those instead?

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: But then we need to iterate twice over them?

Comment (DaughterOfMars) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: that's fair

Comment (thibault-martinez) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: Wait why twice?

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: `implicit_accounts()` iterates over the outputs to filter for implicit account creation outputs and we would need to iterate again and filter for outputs with this address, now both things are done at once

Comment (thibault-martinez) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: You can just keep filtering the Iterator, no need to iterate again?

Comment (thibault-martinez) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: > `implicit_accounts()` iterates over the outputs

It does not! It returns an iterator, bit different

Comment (DaughterOfMars) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: oh true

Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/mod.rs:108]: Changed

**Review by thibault-martinez — APPROVED** [review]

**Review by DaughterOfMars — APPROVED** [review]

**Pull request merged**